Fix history and restart frequency, new features to scripts

apcraig opened the pull request:
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Fix history and restart frequency, update scripts, update unit tests
- [X] Developer(s): 
    apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    All tests are bit-for-bit on cheyenne with full test suite three compilers, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#0cd4402af2528228ea3aa33dc34e426e5a353474
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [X] Please provide any additional information or relevant details below:

    Fix history/restart frequency bugs
    
    - Fix bugs in history/restart frequency associated with new calendar, closes #589
    - Add histfreq_base and dumpfreq_base namelist ('init' or 'zero') to specify reference date for history and restart output.  defaults are 'zero' and 'init' respectively for hist and dump.  Setting histfreq_base to 'zero' allows for consistent output across multiple runs.  Setting dumpfreq_base to 'init' allows the standard testing which requires restarts be written, for example, 5 days after the start of the run 
    - Create compute_relative_elapsed method in ice_calendar to improve code reuse
    - Update set_nml.histall to include hourly output

    Update initial/restart logic

    - "restart" namelist is deprecated, now computed internally
    - Modify initial/continue init checks and set restart and use_restart_time as needed, closes #527

   New features in scripts

    - Update test scripts to cleanly abort if run fails where possible, closes #608
    - Update decomp test so it's rerunable, remove restart at start of run, closes #601
    - Add ability to do bfbcomp tests with additional options set on command line, closes #569

    Update documentation

    - Update documentation of calendar frequency computation, calendar types, and closed boundaries, closes #541
    - Update documentation with regard to initial/continue modes

    Update unit tests

    - Add optional doabort flag to abort_ice to control whether the method aborts.  This is useful for testing and code coverage statistics, although doabort=.false. will not call the actual abort method, but we can test the interfaces and rest of the code.   
    - Clean up some of the unit tests to better support regression testing.  Remove extra abort calls in bcstchk and sumchk on runs that complete fine but don't pass checks.  These aborts should never have been there.

DeniseWorthen commented:
@apcraig I've tested this in ufs-weather;  I can now obtain 6hrly history files as before. Everything is b4b with our current baselines.

apcraig commented:
Thanks @DeniseWorthen!

eclare108213 reviewed (commented):
This generally looks okay to me.   This could be done in a later PR, but I think the documentation needs a bit more information added about how the various calendar and restart options interact with each other.  E.g. the table showing ice_ic, runtype and restart values could be expanded to indicate how the *_init dates play in, depending on use_restart_time (or show some examples).  Are use_restart_time and dumpfreq_base completely independent?

phil-blain commented on cicecore/shared/ice_calendar.F90:
```diff
@@ -126,10 +126,9 @@ module ice_calendar
-                             ! default = ' '
+         dumpfreq_base = 'zero', & ! restart frequency basetime ('zero', 'init')
+         histfreq_base = 'init', & ! history frequency basetime ('zero', 'init')
+         calendar_type             ! define calendar type
```
why are the default values set here in **`ice_calendar`** instead of in **`ice_init`** as for other namelist flags ?

apcraig replied:
They are also set in ice_init in the standard method.  The reason that I set them here as well is to help the unit testing work better.  calchk doesn't go thru any CICE initialization, so having defaults set here means I don't have to explicitly set them when I run the calchk unit test.  One could argue whether that is the best way to do it.  I also made a similar change to ndtd in ice_calendar.

phil-blain commented on cice.setup:
```diff
@@ -786,6 +786,11 @@ EOF
       if ($bfbcomp != ${spval}) then
         set fbfbcomp = ${machcomp}_${bfbcomp_test_grid_pes}${bfbcomp_soptions}
       endif
+    else
```
we are far below the ` if (${docase} == 0) then` line here, so for documentation purpose maybe a comment after the `else` would help to clarify where we are in the logic?

phil-blain replied:
oups, I l left this comment on an earlier commit of the series, the most recent does it differently so it's ok.

apcraig replied:
This has actually changed in a more recent commit.  It's now outside the docase if test and establishes the standard.  It's only with suites that we invoke the more complex implementation.

phil-blain commented on doc/source/cice_index.rst:
```diff
@@ -28,25 +28,25 @@ either Celsius or Kelvin units).
-   "a_rapid_mode", ":math:`{\bullet}` brine channel diameter", ""
-   "add_mpi_barriers", ":math:`\bullet` turns on MPI barriers for communication throttling", ""
-   "advection", ":math:`\bullet` type of advection algorithm used (‘remap’ or ‘upwind’)", "remap"
+   "a_rapid_mode", "brine channel diameter", ""
+   "add_mpi_barriers", "turns on MPI barriers for communication throttling", ""
+   "advection", "type of advection algorithm used (‘remap’ or ‘upwind’)", "remap"
```
Here and below, all the bullets are removed. From what I understand, these were used to denote namelist settings, and now that we have a separate table of namelist options, the bullets are redundant. Is that the reasoning for this change ? Maybe a mention of that rationale in the commit message would be good ?

apcraig replied:
Interesting, I couldn't figure out why those bullets were there.  It just looked like some had them and others didn't which seemed inconsistent to me.  So I just removed them all.   If we need to restore and review all the current variables, we can do that.  Sorry about that.

phil-blain replied:
I think it's ok to delete them, and remove the mention about them at the top of the index, like you do in your latest version. Closing this conversation.

phil-blain replied:
... turns out I can't do that :p It turns out I instead hid my first comment above.

apcraig commented on cicecore/cicedynB/general/ice_init.F90:
```diff
@@ -1116,6 +1121,16 @@ subroutine input_data
          abort_list = trim(abort_list)//":22"
       endif
 
+      if(histfreq_base /= 'init' .and. histfreq_base /= 'zero') then
```
We don't, but I think it's reasonable to trap them during the initialization for efficiency.  I think it's also proper to have the if loop in the calendar operate only on valid values.

phil-blain commented on cicecore/shared/ice_calendar.F90:
```diff
@@ -402,9 +452,31 @@ subroutine calendar()
 
       ! Restart writing flag
 
+      if (dumpfreq_base == 'zero') then
+         elapsed_years = myear
+         elapsed_months =  elapsed_years * months_per_year + mmonth - 1
+         elapsed_days = compute_days_between(0,1,1,myear,mmonth,mday)
+         elapsed_hours = elapsed_days * hours_per_day + hour
+      elseif (dumpfreq_base == 'init') then
+         elapsed_years = myear - year_init
+         elapsed_months =  elapsed_years * months_per_year + (mmonth - month_init)
+         elapsed_days = compute_days_between(year_init,month_init,day_init,myear,mmonth,mday)
+         elapsed_hours = elapsed_days * hours_per_day + hour
+      else
+         write(nu_diag,*) trim(subname),' ERROR dumpfreq_base not recognized, ',trim(dumpfreq_base)
+         call abort_ice(subname//'ERROR: dumpfreq_base value invalid')
+      endif
```
I think it would be cleaner to have a separate subroutine that computes `elapsed_{years,months,days,hours}` depending on the 'init' and 'zero' setting, and have separate variables for the restart and history parts of the subroutine. This way the computation themselves are not repeated in the code.

apcraig replied:
Good point.  I don't know if we need different variables but creating a new subroutine is a good idea.

apcraig replied:
I implemented a new subroutine to improve code reuse.

phil-blain reviewed (commented):
Thanks for these fixes @apcraig. I left a few notes but is mostly looks good. It's nice to see that it was not too hard to add support for both 'zero' and 'init' modes.

apcraig commented on configuration/scripts/ice_in:
```diff
@@ -24,6 +24,7 @@
     pointer_file   = './ice.restart_file'
     dumpfreq       = 'd'
     dumpfreq_n     = 1
+    dumpfreq_base  = 'init'
```
Will do.

apcraig replied:
Done.

apcraig commented on doc/source/user_guide/ug_case_settings.rst:
```diff
@@ -162,6 +162,8 @@ setup_nml
    "", "``m``", "write restart every ``dumpfreq_n`` months", ""
    "", "``y``", "write restart every ``dumpfreq_n`` years", ""
    "", "``1``", "write restart every ``dumpfreq_n`` time step", ""
+   "``dumpfreq_base``", "init", "restart output frequency is relative to the model initial date", "init"
```
This is a very good question.  'init' means relative to year_init, month_init, day_init.  So while those generally do not play a role after the initial run, they now do if freq_base is 'init'.  Is there a reason to support a mode where we create a freq_base relative to the start of the current run.  It probably would not be hard to do that.  It could create chaos in the sense that if you write output every 4 days and run 10 days at a time, you'd have output on day 4, 8, 14, 18 in that mode.  In zero mode you'd have it every 4 days with a possible phase shift (i.e. 1, 5, 9, 13, 17).  And in init mode you'd have it on days 4, 8, 12, 16.  As I alluded to, there are so many things we could do.

Let me know if we need another mode here.

apcraig replied:
@eclare108213 and I decided we didn't need another mode at this time.  Documentation has been updated to clarify.

phil-blain commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -571,7 +578,62 @@ to set the initial model time to the namelist values defined by
+``zero`` and ``init``.  If set to ``zero``, all output will be relative 
+to the absolute reference date, 0000-01-01-00000.  This is the default
+value for ``histfreq_base``, so runs with different initial
+dates will have identical output.  If set to ``init``, all frequencies
+will be relative to the model init time specified by ``year_init``,
+``month_init``, and ``day_init``.  This is the default for
```
Maybe we would want to emphasize specifically that `sec_init` is _not_ taken into account ?

apcraig replied:
Done.

apcraig commented:
I have updated the implementation.  I fixed several issues and also updated the initial/restart implementation after discussions with @eclare108213.  The documentation was also update to reflect these changes.

apcraig commented:
Also, I am running a full test suite on cheyenne now.

apcraig commented on cicecore/shared/ice_calendar.F90:
```diff
@@ -122,14 +126,18 @@ module ice_calendar
          hour         ! hour of the day
 
+      integer (kind=int_kind), parameter :: &
+         myear_max = 200000           ! maximum year, limited by integer overflow in elapsed_horus
```
Fixed.

phil-blain commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -684,18 +687,65 @@ layers and the ice thickness distribution defined by ``kcatbound`` = 0.
+   | `initial`      | `default`                | not used                             | latitude dependent internal ic,        |
+   |                |                          |                                      | namelist defines start date            |
+   +----------------+--------------------------+--------------------------------------+----------------------------------------+
+   | `initial`      | *filename*               | false                                | read file, namelist defines start date |
+   +----------------+--------------------------+--------------------------------------+----------------------------------------+
+   | `initial`      | *filename*               | true                                 | read file, file defines start date     |
```
Maybe be a little more verbose ?

```suggestion
   | `initial`      | *filename*               | false                                | read ice state from file, namelist defines start date |
   +----------------+--------------------------+--------------------------------------+----------------------------------------+
   | `initial`      | *filename*               | true                                 | read file state from file, file defines start date     |
```

apcraig replied:
Done.

phil-blain commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -684,18 +687,65 @@ layers and the ice thickness distribution defined by ``kcatbound`` = 0.
+and ``sec_init``.  These combinations options are summarized in 
+:ref:`tab-ic`. 
+
+Restart files and initial condition files are generally the same file.
```
I'm not sure about the formulation here. Maybe "are generally in the same format" ?

apcraig replied:
I updated this.

phil-blain commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -684,18 +687,65 @@ layers and the ice thickness distribution defined by ``kcatbound`` = 0.
+Three namelist variables generally control model initialization, ``runtype``,
+``ice_ic``, and ``use_restart_time``.  The valid values for ``runtype``
+are ``initial`` or ``continue``.  When ``runtype`` = `continue`, the
+initial condition is defined by the rpointer file, ``use_restart_time``
```
I'm not sure I like the name "rpointer". I would prefer something like "ice pointer file" or something of sorts... and maybe add a "(see below)" because it's the first time the name is introduced.

apcraig replied:
rpointer is a term in CESM.  I have changed those to "pointer".  thanks for catching that.

phil-blain commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -684,18 +687,65 @@ layers and the ice thickness distribution defined by ``kcatbound`` = 0.
+By having this feature, the ice namelist does not need to be constantly
+updated with the latest
+restart filename, and the model can be automatically resubmitted.
+Manually editing the rpointer file in the middle of a run will reset
+the restart filename, and this makes it possible to relatively easily
+backup a run to rerun.
```
I'm not sure exactly what you mean here... also I guess you meant to write "back up" and not "backup" ?

apcraig replied:
I have simplified that sentence.

phil-blain reviewed (commented):
Thanks for these additional clean-ups Tony, this was really long overdue I think. Especially the "Ice initialization" table in the doc, the new version is really clearer. I think it makes sense to start deprecating the `restart` namelist flag,  since it was reset already internally under several conditions.

phil-blain reviewed (approved):
Thanks for these fixes @apcraig. I left a few notes but is mostly looks good. It's nice to see that it was not too hard to add support for both 'zero' and 'init' modes.

phil-blain commented:
@apcraig @eclare108213 @dabail10 :

I remembered this issue: https://github.com/CICE-Consortium/CICE/issues/527. I just skimmed it and from what I understand it would be closed by this here PR with the latest changes to the restart flags, right ? If so we could also check it in the list at https://github.com/CICE-Consortium/CICE/issues/567.

apcraig commented:
I think this does address #527.  In general, we could support use_restart_time=.false. with runtype='continue' and were up to now.  But after some discussion, it was decided to simplify the options a bit, so we're eliminating that feature (for now).  The issue in #527 could (should) have also been solved by modifications in the nuopc cap or changes to the namelist generation in CESM, and maybe that was done or is still worth doing.

Unless other disagree, I think this can close 527 and will add that comment in the PR documentation.

apcraig commented:
I tested the latest changes and everything still works.  This could be merged unless there are other comments to address.

apcraig commented:
Unless I hear any concerns or someone gets to it before me, I'll merge this PR in the next couple days.  Thanks.

phil-blain commented:
Thanks a lot for these fixes @apcraig ! The history frequency bug was the last missing bit for me to wrap up the first phase of our project for updating to CICE6 in our forecasting systems!
